addons/sale/models/sale_order.py

```diff
@@ -906,22 +915,48 @@ def _action_confirm(self):
                     order._create_analytic_account()
 
     def _send_order_confirmation_mail(self):
-        if not self:
+        """ Send a mail to the customer each time an order is confirmed.
+
+        :return: None
+        """
+
+        for sale_order in self:
+            mail_template = sale_order._get_confirmation_template()
+            sale_order._send_order_notification_mail(mail_template)
+
+    def _send_payment_succeeded_for_order_mail(self):
+        """ Send a mail to the customer each time a payment is executed.
+
+        :return: None
+        """
+        for sale_order in self:
+            mail_template = self.env.ref(
+                'sale.mail_template_sale_payment_executed', raise_if_not_found=False
+            )
```

Comment:
The `ref` is independent of the orders, must be moved outside the loop (ref calls cost one SQL query, you are making len(self) calls for 'nothing' here)